pkg/executor/executor.go

```diff
@@ -0,0 +1,126 @@
+package handler
+
+import (
+	"bytes"
+	"github.com/go-logr/logr"
+	"io/ioutil"
+	"strconv"
+
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+)
+
+type executorHandler struct {
+	log         logr.Logger
+	preprocess  string
+	predictor   string
+	postprocess string
+	transport   http.RoundTripper
+}
+
+func New(log logr.Logger, preprocess, predictor, postprocess string) http.Handler {
+	return &executorHandler{
+		log:         log,
+		preprocess:  preprocess,
+		predictor:   predictor,
+		postprocess: postprocess,
+	}
+}
+
+// Return a postProcessor function which can be called by the rever proxy to change the response.
+func (eh *executorHandler) createPostProcessor() func(response *http.Response) error {
+
+	f := func(resp *http.Response) (err error) {
+
+		eh.log.Info("Calling post-processor")
+
+		b, err := ioutil.ReadAll(resp.Body) //Read html
+		if err != nil {
+			return err
+		}
+		err = resp.Body.Close()
+		if err != nil {
+			return err
+		}
+
+		target := &url.URL{
+			Scheme: "http",
+			Host:   eh.postprocess,
+		}
+		reader := bytes.NewReader(b)
+		eh.log.Info("Calling postprocessor", "url", target.String())
+		respPost, err := http.Post(target.String(), resp.Header.Get("Content-Type"), reader)
+
+		b, err = ioutil.ReadAll(respPost.Body) //Read html
+		if err != nil {
+			return err
+		}
+		err = respPost.Body.Close()
+		if err != nil {
+			return err
+		}
+
+		body := ioutil.NopCloser(bytes.NewReader(b))
+		resp.Body = body
+		resp.ContentLength = int64(len(b))
+		resp.Header.Set("Content-Length", strconv.Itoa(len(b)))
+		return nil
+	}
+	return f
+}
+
+// Call a preprocessor and change the incoming request with its response
+func (eh *executorHandler) preProcess(r *http.Request) error {
+	target := &url.URL{
+		Scheme: "http",
+		Host:   eh.preprocess,
+	}
+	eh.log.Info("Calling preprocessor", "url", target.String())
+	respPost, err := http.Post(target.String(), r.Header.Get("Content-Type"), r.Body)
+
+	b, err := ioutil.ReadAll(respPost.Body) //Read html
+	if err != nil {
+		return err
+	}
+	err = respPost.Body.Close()
+	if err != nil {
+		return err
+	}
+
+	r.Body = ioutil.NopCloser(bytes.NewReader(b))
+	r.ContentLength = int64(len(b))
+
+	return nil
+}
+
+// Create a reverse proxy to call the predictor with optional pre and post processing
+func (eh *executorHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	if len(eh.preprocess) > 0 {
+		err := eh.preProcess(r)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	target := &url.URL{
+		Scheme: "http",
+		Host:   eh.predictor,
+	}
+
+	r.URL.Host = target.Host
+	r.URL.Scheme = target.Scheme
+	r.Header.Set("X-Forwarded-Host", r.Header.Get("Host"))
+	r.Host = eh.predictor
+
+	eh.log.Info("About to proxy request", "host", eh.predictor)
+
+	proxy := httputil.NewSingleHostReverseProxy(target)
```

Comment:
It depends if when the executor is added there is still an exposed endpoint for the predictor that allows users to bypass the executor and what each endpoint is called, i.e. if the endpoint for the executor takes on the name of your KFService as the main endpoint?

As for technology, it seemed a useful package to use to forward the core request on to the predictor. Do you have objections to using `httputil` code for this operation?

Reply:
I think they probably should take on the same name.
We may also want to alias the predictor as: 
predictor.<model>.domain

Reply:
So no changes here then @ellis-bigelow ?

Reply:
Hey! I haven't had the time to deeply investigate. I definitely lean away from the proxy approach, but I don't have solid reasons why. I think the main idea (as stated above) is that I see a proxy as something that's transparently placed in front of a server and simply forwards traffic, not fulfilling the level of complexity we want this to do. 

For pre/pred/post the approach is cleanish, but I think it might be a little less flexible as we incorporate more features.

Reply:
Updated to remove proxy.